DotVVM.Diagnostics.StatusPage/DotHtmlFilesRuntimePrecompiler.cs

```diff
@@ -0,0 +1,180 @@
+
+                    if (compiledControl is DotvvmView view && view.Directives.TryGetValue(
+                            ParserConstants.MasterPageDirective,
+                            out var masterPage))
+                    {
+                        if (MasterPages.All(s => s.VirtualPath != masterPage) &&
+                            tempList.All(s => s.VirtualPath != masterPage))
+                        {
+                            tempList.Add(new DotHtmlFileInfo()
+                            {
+                                VirtualPath = masterPage
+                            });
+                        }
+                    }
+
+                    file.Status = CompilationState.CompletedSuccessfully;
+                    file.Exception = null;
+                }
+                catch (Exception e)
+                {
+                    file.Status = CompilationState.CompilationFailed;
+                    file.Exception = e.Message;
+                }
+            }
+        }
+        /// <summary>
+        /// Builds all dothtml file registered in configuration. When the files are already compiled and cached, compilation is not executed.  
+        /// </summary>
+        public async Task<bool> CompileAll()
+        {
+            ResolveAllFiles();
+            var tempMasterPages = new ConcurrentBag<DotHtmlFileInfo>();
+            await BuildPagesAndControls(tempMasterPages);
+            await BuildMasterPages(tempMasterPages);
+            return MasterPages.Concat(Controls).Concat(Routes).All(s => s.Status != CompilationState.CompilationFailed);
+        }
+
+        private async Task<ConcurrentBag<DotHtmlFileInfo>> BuildMasterPages(ConcurrentBag<DotHtmlFileInfo> tempMasterPages)
+        {
+            while (tempMasterPages.Count > 0)
+            {
+                tempMasterPages.ToList().ForEach(masterPages.Add);
+                masterPages = new ConcurrentBag<DotHtmlFileInfo>(masterPages.Distinct());
+                tempMasterPages = new ConcurrentBag<DotHtmlFileInfo>();
+
+                if (statusPageOptions.BuildInParallel)
+                {
+                    await Task.WhenAll(masterPages.Select(i => Task.Run(() => BuildView(i, tempMasterPages))).ToArray());
+                }
+                else
+                {
+                    foreach (var item in masterPages)
+                    {
+                        BuildView(item, tempMasterPages);
+                    }
+                }
+            }
+
+            return tempMasterPages;
+        }
+
+        private async Task BuildPagesAndControls(ConcurrentBag<DotHtmlFileInfo> tempMasterPages)
+        {
+            if (statusPageOptions.BuildInParallel)
+            {
+                var compileTasks = Routes.Select(a => Task.Run(() => BuildView(a, tempMasterPages))).ToList();
+                compileTasks.AddRange(Controls.Select(a => Task.Run(() => BuildView(a, tempMasterPages))).ToList());
```

Comment:
The ToList is redundant here. I'd also prefer to use `Enumerable.Concat` in this simple case.